### PR TITLE
Add social buttons to sign up

### PR DESCRIPTION
**Pull request by Tug**
This PR adds 2 buttons for social sign up on WordPress.com in the `user` step of signup (enabled on the new `social` flow only for now).
We still need to implement the login endpoints on wpcom for them to be operational.

image

**(Outdated)**: 
> This reuses 2 components: [react-facebook-login-component](https://github.com/kennetpostigo/react-facebook-login-component) and [react-google-login-component](https://github.com/kennetpostigo/react-google-login-component) both of which are not really suited for our needs, so we will need to reimplement them (or make a contribution).
> 
> - In the case of `react-google-login-component` it seems the popup is blocked by the browser the first time, this is probably due to [this call](https://github.com/kennetpostigo/react-google-login-component/blob/master/src/GoogleLogin.js#L33) being asynchronous (so we should preload the library).
> - In the case of `react-facebook-login-component`, we don't want to call facebook's endpoint `/me` from the client, we just need to receive the access_token to do it from the server.

### Testing Instructions
- Run this locally and go to http://calypso.localhost:3000/start/social or open https://calypso.live/start/social?branch=add/social-signup-flow (need to be logged out)
- Check that the styles match the prototype
- Click on each social button and verify that a popup opens and display the oauth authorization page.
- Look at the console and check that we receive a response with an `access_token` each time

### Reviews
- [x] Code
- [x] Product

### CODE REVIEW

File: client/signup/config/steps.js [hunk 1]
```diff
@@ -143,4 +143,13 @@ module.exports = {
 		providesDependencies: [ 'siteId', 'siteSlug', 'domainItem', 'themeSlugWithRepo' ],
 		delayApiRequestUntilComplete: true
 	},
+
+	'user-social': {
+		stepName: 'user-social',
+		props: {
+			headerText: i18n.translate( "Let's get started" ),
+			subHeaderText: i18n.translate( "Welcome to the wonderful WordPress.com community." )
```
Comment (a8ci18n) [inline client/signup/config/steps.js]: Hi! I've found a possible matching string that has already been translated **20** times:
`translate( 'Welcome to the wonderful WordPress.com community' )` *ES Score: 21*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/signup/config/steps.js [hunk 2]
```diff
@@ -143,4 +143,13 @@ module.exports = {
 		providesDependencies: [ 'siteId', 'siteSlug', 'domainItem', 'themeSlugWithRepo' ],
 		delayApiRequestUntilComplete: true
 	},
+
+	'user-social': {
+		stepName: 'user-social',
+		props: {
+			headerText: i18n.translate( "Let's get started" ),
```
Comment (a8ci18n) [inline client/signup/config/steps.js]: Hi! I've found a possible matching string that has already been translated **32** times:
`translate( 'Let's get started.' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/components/signup-form/index.jsx [hunk 3]
```diff
@@ -447,6 +456,10 @@ export default React.createClass( {
 							{ this.props.formHeader }
 						</header>
 					}
+					{ this.props.isSocialSignupEnabled && <SocialSignupForm /> }
+					{ this.props.isSocialSignupEnabled && <HrWithText>
+						{ i18n.translate( 'Or sign up with your email address:' ) }
```
Comment (a8ci18n) [inline client/components/signup-form/index.jsx]: Hi! I've found a possible matching string that has already been translated **15** times:
`translate( 'Enter your sign up email address' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: package.json [hunk 4]
```diff
@@ -106,6 +106,8 @@
     "react-day-picker": "2.4.1",
     "react-docgen": "2.13.0",
     "react-dom": "15.4.0",
+    "react-facebook-login-component": "0.6.0",
```
Comment (drewblaisdell) [inline package.json]: If we're going to replace these with our own buttons, I don't think we should import them into this project at all. What do you think about just using placeholder `<Button>` elements right now?

File: client/components/signup-form/social/style.scss [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/*rtl:ignore*/
```
Comment (drewblaisdell) [inline client/components/signup-form/social/style.scss]: Per [these docs](https://github.com/Automattic/wp-calypso/blob/master/docs/components.md), I think the component for these social buttons should either be a standalone component (in `/client/components`) or a sub-component in the same directory as `SignupForm`. Then, instead of styles like `.signup-form__social__button` we would either have:

- `.social-buttons__button` if we created a standalone component named `SocialButtons`.
- `.signup-form__social-button` if we created a sub-component.

File: client/components/hr-with-text/style.scss [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// ==========================================================================
+// .hr-with-text
+//
+// An hr-like line with text on top:
```
Comment (drewblaisdell) [inline client/components/hr-with-text/style.scss]: Good idea to create a separate component for this.

File: client/signup/steps/user/index.jsx [hunk 7]
```diff
@@ -149,6 +150,7 @@ export class UserStep extends Component {
 				submitForm={ this.submitForm }
 				submitButtonText={ this.submitButtonText() }
 				suggestedUsername={ this.props.suggestedUsername }
+				isSocialSignupEnabled={ isEnabled( 'signup/social' ) }
```
Comment (drewblaisdell) [inline client/signup/steps/user/index.jsx]: Instead of adding this by default to every user step in development, it would probably be a little bit less disruptive to create a separate, development-only flow for testing social signup with a version of this step with `isSocialSignupEnabled` set to true. This way, there isn't any added confusion about broken signup buttons in `/start` in development. What do you think?

File: package.json (hunk 4 above)
Comment (Tug) [inline package.json]: I just rewrote them

File: client/components/signup-form/social/style.scss (hunk 5 above)
Comment (Tug) [inline client/components/signup-form/social/style.scss]: Refactored a bit all this, tell me what you think

File: client/signup/steps/user/index.jsx (hunk 7 above)
Comment (Tug) [inline client/signup/steps/user/index.jsx]: Done!

**Comment by Tug** [comment]
@drewblaisdell Thanks you for reviewing, I ended up rewriting those social button components and addressed the other concerns. Tell me what you think!

File: client/signup/steps/user/index.jsx [hunk 8]
```diff
@@ -149,6 +150,7 @@ export class UserStep extends Component {
 				submitForm={ this.submitForm }
 				submitButtonText={ this.submitButtonText() }
 				suggestedUsername={ this.props.suggestedUsername }
+				isSocialSignupEnabled={ isEnabled( 'signup/social' ) && this.props.flowName === 'social' }
```
Comment (scruffian) [inline client/signup/steps/user/index.jsx]: I think it would be better to create a different step that implements the same component and then pass prop down, rather than rely on the flow name here.

Comment (Tug) [inline client/signup/steps/user/index.jsx]: We could do that, I wanted to avoid making the PR larger though.
Question: is this `social` flow going to be used in production or is this just for us?

**Comment by scruffian** [comment]
This looks good. What do you think about making the Google and Facebook buttons  a separate repo?

File: client/signup/steps/user/index.jsx (hunk 8 above)
Comment (scruffian) [inline client/signup/steps/user/index.jsx]: I don't imagine the flow being used in production.

**Comment by Tug** [comment]
Thanks for your review @scruffian 

> What do you think about making the Google and Facebook buttons a separate repo?
We could but they are rather simple now and they depend on their styles so they would not be that much useful for other project as they are.

We would need to update the `render` functions from this:
```
<button className="button" onClick={ this.handleClick }>
	<span>
		<SocialLogo className="social-buttons__logo" icon="google" size={ 24 } />
		<span className="social-buttons__service-name">
			{ this.props.text || 'Google' }
		</span>
	</span>
</button>
```
to this:
```
<button className="button" onClick={ this.handleClick }>
	{ this.props.children }
</button>
```
to make them more modular.

I see little gain for now. I say we wait until we have them working completely in calypso then we reconsider. What do you say?

File: client/signup/steps/user/index.jsx (hunk 8 above)
Comment (Tug) [inline client/signup/steps/user/index.jsx]: `UserStep` is a bit large so I'd rather not duplicate it just for our tests (also I will risk making this PR unmergeable because of its size).

**Comment by scruffian** [comment]
> I say we wait until we have them working completely in calypso then we can reconsider. What do you say?

If you think that's better than that's fine with me :)

File: client/signup/steps/user/index.jsx (hunk 8 above)
Comment (scruffian) [inline client/signup/steps/user/index.jsx]: It's not necessary to duplicate the whole step component. Different steps can share the same component, for example: https://github.com/Automattic/wp-calypso/blob/16d8d9c5388c507eb8dd7c6d8988793dfc4609fa/client/signup/config/steps.js#L92

Comment (Tug) [inline client/signup/steps/user/index.jsx]: Oh cool, I didn't know

File: client/components/social-buttons/facebook.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import SocialLogo from 'social-logos';
+
+export default class FacebookLoginButton extends React.Component {
```
Comment (drewblaisdell) [inline client/components/social-buttons/facebook.js]: We should probably add `propTypes` to this component.

File: client/components/social-buttons/facebook.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+				resolve( window.FB );
+			} else {
+				( function ( d, s, id ) {
+					const element = d.getElementsByTagName( s )[0];
+					const fjs = element;
+					let js = element;
+					if ( d.getElementById( id ) ) { return; }
+					js = d.createElement( s ); js.id = id;
+					js.src = '//connect.facebook.net/en_US/sdk.js';
+					fjs.parentNode.insertBefore( js, fjs );
```
Comment (drewblaisdell) [inline client/components/social-buttons/facebook.js]: In the past, we've used [`loadScript`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/load-script/index.js#L13) for loading third party scripts like this, as it is a bit clearer than the script Facebook asks us to copy/paste into the app. What do you think about using it here?

File: client/components/social-buttons/facebook.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+					resolve( window.FB );
+					delete window.fbAsyncInit;
+				};
+			}
+		} );
+	}
+
+	handleClick() {
+		this.state.FB.login( response => {
+			if ( this.props.responseHandler ) {
```
Comment (drewblaisdell) [inline client/components/social-buttons/facebook.js]: I think `responseHandler` should be a required prop, so that we don't need a presence check here.

File: client/components/social-buttons/facebook.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+				<button className={ this.props.className }>&nbsp;</button>
+			);
+		}
+
+		return (
+			<button className="button signup-form__social__button" onClick={ this.handleClick }>
+				<span>
+					<SocialLogo className="social-buttons__logo" icon="facebook" size={ 24 } />
+					<span className="social-buttons__service-name">
+						{ this.props.facebook || 'Facebook' }
```
Comment (drewblaisdell) [inline client/components/social-buttons/facebook.js]: I don't think we need `this.props.facebook ||` here, as we don't need to display anything other than `Facebook` at this time.

File: client/components/social-buttons/google.js [hunk 13]
```diff
@@ -0,0 +1,75 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import SocialLogo from 'social-logos';
+
+export default class GoogleLoginButton extends React.Component{
```
Comment (drewblaisdell) [inline client/components/social-buttons/google.js]: Same note here regarding `propTypes`.

File: client/components/social-buttons/facebook.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+		this.state.FB.login( response => {
+			if ( this.props.responseHandler ) {
+				this.props.responseHandler( response );
+			}
+		}, { scope: this.props.scope } );
+	};
+
+	render() {
+		// Show an empty button if Facebook Connect API is not loaded
+		if ( ! this.state.FB ) {
```
Comment (drewblaisdell) [inline client/components/social-buttons/facebook.js]: Super minor, but instead of a presence check for a reference to `window.FB`, what do you think about setting a boolean like `isLoaded`? This might be a little clearer, and is the minimum state needed to represent the render states here (as there are only two).

File: client/components/social-buttons/google.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+		const { responseHandler } = this.props;
+
+		this.state.auth2.signIn().then( googleUser => {
+			responseHandler( googleUser );
+		} );
+	}
+
+	render() {
+		// Show an empty button if Google auth2 API is not loaded
+		if ( ! this.state.auth2 ) {
```
Comment (drewblaisdell) [inline client/components/social-buttons/google.js]: Same question about using an `isLoaded` boolean.

File: client/components/social-buttons/google.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+				<button className="button">&nbsp;</button>
+			);
+		}
+
+		return (
+			<button className="button" onClick={ this.handleClick }>
+				<span>
+					<SocialLogo className="social-buttons__logo" icon="google" size={ 24 } />
+					<span className="social-buttons__service-name">
+						{ this.props.text || 'Google' }
```
Comment (drewblaisdell) [inline client/components/social-buttons/google.js]: I don't think we need the `text` prop here.

File: client/components/social-buttons/google.js [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+		this.state.auth2.signIn().then( googleUser => {
+			responseHandler( googleUser );
+		} );
+	}
+
+	render() {
+		// Show an empty button if Google auth2 API is not loaded
+		if ( ! this.state.auth2 ) {
+			return (
+				<button className="button">&nbsp;</button>
```
Comment (drewblaisdell) [inline client/components/social-buttons/google.js]: I think we probably want to tweak the loading state here, as currently a single blank button is displayed (when both are loading):

screen shot 2017-03-16 at 12 36 35 pm

I think, ideally, we would allow users to click this button early and simply call `auth2.signIn()` once the script finished loading, but I think just improving the loading state so that it says `Loading…` next to the logo would be sufficient for now.

File: client/components/social-buttons/facebook.js (hunk 14 above)
Comment (Tug) [inline client/components/social-buttons/facebook.js]: I decided to render the component even if the library is not loaded. If the user clicks on the button before it's loaded we try to open the popeup asynchronously (we don't have much choice anyway).

File: client/components/social-buttons/google.js (hunk 17 above)
Comment (Tug) [inline client/components/social-buttons/google.js]: See https://github.com/Automattic/wp-calypso/pull/12115#discussion_r106626157

**Comment by Tug** [comment]
Should be good now

File: client/components/social-buttons/facebook.js [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+
+	// See: https://developers.facebook.com/docs/javascript/reference/FB.init/v2.8
+	static propTypes = {
+		appId: PropTypes.string.isRequired,
+		version: PropTypes.string,
+		cookie: PropTypes.bool,
+		status: PropTypes.bool,
+		xfbml: PropTypes.bool,
+		responseHandler: PropTypes.func.isRequired,
+	};
```
Comment (stephanethomas) [inline client/components/social-buttons/facebook.js]: We're missing a prop type for `scope`.

File: client/components/hr-with-text/index.jsx [hunk 19]
```diff
@@ -0,0 +1,12 @@
...
+ */
+import React from 'react';
+
+export default ( { children } ) => (
+	<div className="hr-with-text">
+		<div>
+			{ children }
+		</div>
+	</div>
+);
```
Comment (Tug) [inline client/components/hr-with-text/index.jsx]: What propTypes? `children` is a special kind of prop.

File: client/components/signup-form/index.jsx [hunk 20]
```diff
@@ -447,6 +456,10 @@ export default React.createClass( {
...
+					</HrWithText> }
```
Comment (Tug) [inline client/components/signup-form/index.jsx]: I think it's mostly equivalent, not more nor less readable

File: client/components/signup-form/social.jsx [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+		}
+		// TODO: post response to the new wpcom endpoint to login
+	}
+
+	onFacebookResponse( response ) {
+		if ( ! response.email ) {
+			return;
+		}
+		// TODO: post response to the new wpcom endpoint to login
+	}
```
Comment (Tug) [inline client/components/signup-form/social.jsx]: It's not "authenticating with google" though, it's the result of the authentication, after the flow is over. We will need to call the wpcom endpoint to authenticate using google `id_token` and we might indeed name the function called here `authenticateWithGoogle`.

**Comment by Tug** [comment]
Thanks for the review @stephanethomas. I believe i addressed most of your concerns.

File: client/components/hr-with-text/index.jsx (hunk 19 above)
Comment (stephanethomas) [inline client/components/hr-with-text/index.jsx]: Yes, I was indeed talking about `children` here (a `PropTypes.node`).

Comment (Tug) [inline client/components/hr-with-text/index.jsx]: You were right it seems we can document the `children` prop this way.

File: client/components/social-buttons/facebook.js [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+			} );
+		} );
+	}
+
+	handleClick() {
+		const { responseHandler, scope } = this.props;
+
+		// Handle click async if the library is not loaded yet
+		// the popup might be blocked by the browser in that case
+		this.loadFacebookAuth().then( () => {
```
Comment (drewblaisdell) [inline client/components/social-buttons/facebook.js]: Instead of creating a new promise in this case, could we return the existing promise that was created in `componentWillMount`? Otherwise, we'll be calling `loadScript` twice and (I believe) injecting an extra `<script>` into the header.

File: client/components/social-buttons/facebook.js [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+		version: 'v2.8',
+		cookie: false,
+		status: false,
+		xfbml: true,
+		scope: 'public_profile,email',
+	};
+
+	constructor( props ) {
+		super( props );
+		this.FB = null;
```
Comment (drewblaisdell) [inline client/components/social-buttons/facebook.js]: Super minor, but why use `this.FB` at all? The script populates `window.FB`, which we already reference, so it seems like it muddies the water a bit to access the same object twice through two different references.

File: client/components/social-buttons/google.js [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+	};
+
+	constructor( props ) {
+		super( props );
+		this.googleAuth = null;
+		this.handleClick = this.handleClick.bind( this );
+	}
+
+	componentWillMount() {
+		this.loadGoogleAuth();
```
Comment (yurynix) [inline client/components/social-buttons/google.js]: By not storing the promise here, we'll initiate a second request in `handleClick`

Comment (yurynix) [inline client/components/social-buttons/google.js]: Ahh, I see, `this.googleAuth` suppose to handle that, never mind

File: client/components/social-buttons/google.js [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+		}
+
+		return new Promise( resolve => {
+			loadScript( 'https://apis.google.com/js/platform.js', () => resolve( window.gapi ) );
+		} );
+	}
+
+	loadGoogleAuth() {
+		return new Promise( resolve => {
+			if ( this.googleAuth ) {
```
Comment (yurynix) [inline client/components/social-buttons/google.js]: The problem with that approach:
```js
this.loadGoogleAuth().then( () => conosole.log( '1' ) );
this.loadGoogleAuth().then( () => conosole.log( '2' ) );
this.loadGoogleAuth().then( () => conosole.log( '3' ) );
```

This code will initiate 3 load requests, because we're setting `this.googleAuth` only after we finished loading,
we should "lock" when we start loading.

**Comment by fabianapsimoes** [comment]
> Click on each social button and verify that a popup opens and display the oauth authorization page.

I get a popup for Google, but not for Facebook. Is that expected?

It also seems that validation is being triggered when the social buttons are clicked. We should prevent that, since people using the social buttons won't be filling that signup form.

**Comment by stephanethomas** [comment]
> It also seems that validation is being triggered when the social buttons are clicked. We should prevent that, since people using the social buttons won't be filling that signup form.

I've rebased this pull request, and fixed that issue (as well as a warning that would appear in the browser's console in development).

> I get a popup for Google, but not for Facebook. Is that expected?

@fabianapsimoes, you should get a popup for Facebook as well. This works for me. Could you give it another try?

File: client/components/social-buttons/google.js [hunk 26]
```diff
@@ -34,39 +34,37 @@ export default class GoogleLoginButton extends Component {
 		}
 
 		return new Promise( resolve => {
-			loadScript( 'https://apis.google.com/js/platform.js', () => resolve( window.gapi ) );
+			loadScript( 'https://apis.google.com/js/api.js', () => resolve( window.gapi ) );
```
Comment (stephanethomas) [inline client/components/social-buttons/google.js]: Just curious, why did you change that @yurynix?

Comment (drewblaisdell) [inline client/components/social-buttons/google.js]: Both allow us to authenticate the user with their Google account, but `platform.js` is larger:

```bash
~ ☯ curl -s https://apis.google.com/js/platform.js | wc -m
   40156
~ ☯ curl -s https://apis.google.com/js/api.js | wc -m
   12313
```

**Comment by drewblaisdell** [comment]
Code :+1: I was also able to authenticate with Google and Facebook properly.

File: client/components/social-buttons/google.js (hunk 26 above)
Comment (yurynix) [inline client/components/social-buttons/google.js]: I changed it because I believed that platform.js is an older version of the API, but I might be mistaken.

I believe that `api.js` provides promises as return value from it's functions such as `gapi.client.init` while `platform.js` uses callback style.

I was using this documentation:
https://developers.google.com/api-client-library/javascript/start/start-js

Haven't thought about checking the size, thank you @drewblaisdell for catching that! =)

File: client/components/social-buttons/facebook.js [hunk 27]
```diff
@@ -75,7 +77,7 @@ export default class FacebookLoginButton extends Component {
 
 		// Handle click async if the library is not loaded yet
 		// the popup might be blocked by the browser in that case
-		this.loadFacebookAuth().then( FB => {
+		this.initialize().then( FB => {
```
Comment (yurynix) [inline client/components/social-buttons/facebook.js]: That's probably a nitpick, but I don't really like the `this.intialize()` name,
I would prefer something like `this.getFacebookApi().then( api => api.login( ... ) )`
something like `getFacebookApi` makes more sense to me than the generic `initialize`

Comment (stephanethomas) [inline client/components/social-buttons/facebook.js]: Good point, this is much better indeed!

**Comment by fabianapsimoes** [comment]
Product :+1:
